File: C#/MyProjectName/session.cs
// Author: Clayton
// Rollno: 31
// Title: Goa Legislative Assembly
// Startdate: 15-07-24
// Modifieddate: 22-07-24
// Description: To show OOP implementation in Goa Legislative Assembly

using System;
using System.Collections.Generic;

public class Session
{
    private string sessionDate;
    private int startHour;
    private int startMinute;
    private int endHour;
    private int endMinute;
    private string agenda;
    private List<string> topicsOfDiscussion;

    public Session(string sessionDate, int startHour, int startMinute, int endHour, int endMinute, string agenda)
    {
        this.sessionDate = sessionDate;
        SetStartHour(startHour);
        SetStartMinute(startMinute);
        SetEndHour(endHour);
        SetEndMinute(endMinute);
        this.agenda = agenda;
        this.topicsOfDiscussion = new List<string>();
    }

    public void SetStartHour(int startHour)
    {
        if (startHour < 0 || startHour > 23)
        {
            throw new ArgumentException("Hour must be between 0 and 23.");
        }
        this.startHour = startHour;
    }

    public void SetStartMinute(int startMinute)
    {
        if (startMinute < 0 || startMinute > 59)
        {
            throw new ArgumentException("Minute must be between 0 and 59.");
        }
        this.startMinute = startMinute;
    }

    public void SetEndHour(int endHour)
    {
        if (endHour < 0 || endHour > 23)
        {
            throw new ArgumentException("Hour must be between 0 and 23.");
        }
        this.endHour = endHour;
    }

    public void SetEndMinute(int endMinute)
    {
        if (endMinute < 0 || endMinute > 59)
        {
            throw new ArgumentException("Minute must be between 0 and 59.");
        }
        this.endMinute = endMinute;
    }

    public void AddTopicOfDiscussion(string topic)
    {
        topicsOfDiscussion.Add(topic);
    }

    public string Agenda
    {
        get { return agenda; }
    }

    public void DisplayInfo()
    {
        Console.WriteLine("Session Date: " + sessionDate);
        Console.WriteLine($"Start Time: {startHour:D2}:{startMinute:D2}");
        Console.WriteLine($"End Time: {endHour:D2}:{endMinute:D2}");
        Console.WriteLine("Agenda: " + agenda);
        Console.WriteLine("Topics of Discussion:");
        foreach (string topic in topicsOfDiscussion)
        {
            Console.WriteLine(" - " + topic);
        }
    }
}

File: C++/session.cpp
#include "Session.h"
#include <iostream>
#include <stdexcept>

// Constructor
Session::Session(const std::string& sessionDate, int startHour, int startMinute, int endHour, int endMinute, const std::string& agenda)
    : sessionDate(sessionDate), agenda(agenda) {
    setStartHour(startHour);
    setStartMinute(startMinute);
    setEndHour(endHour);
    setEndMinute(endMinute);
}

// Helper methods for validation
void Session::validateHour(int hour) const {
    if (hour < 0 || hour > 23) {
        throw std::invalid_argument("Hour must be between 0 and 23.");
    }
}

void Session::validateMinute(int minute) const {
    if (minute < 0 || minute > 59) {
        throw std::invalid_argument("Minute must be between 0 and 59.");
    }
}

// Setters
void Session::setStartHour(int startHour) {
    validateHour(startHour);
    this->startHour = startHour;
}

void Session::setStartMinute(int startMinute) {
    validateMinute(startMinute);
    this->startMinute = startMinute;
}

void Session::setEndHour(int endHour) {
    validateHour(endHour);
    this->endHour = endHour;
}

void Session::setEndMinute(int endMinute) {
    validateMinute(endMinute);
    this->endMinute = endMinute;
}

// Add a topic of discussion
void Session::addTopicOfDiscussion(const std::string& topic) {
    topicsOfDiscussion.push_back(topic);
}

// Getter
std::string Session::getAgenda() const {
    return agenda;
}

// Display information
void Session::displayInfo() const {
    std::cout << "Session Date: " << sessionDate << std::endl;
    std::cout << "Start Time: " << (startHour < 10 ? "0" : "") << startHour << ":" << (startMinute < 10 ? "0" : "") << startMinute << std::endl;
    std::cout << "End Time: " << (endHour < 10 ? "0" : "") << endHour << ":" << (endMinute < 10 ? "0" : "") << endMinute << std::endl;
    std::cout << "Agenda: " << agenda << std::endl;
    std::cout << "Topics of Discussion:" << std::endl;
    for (const auto& topic : topicsOfDiscussion) {
        std::cout << " - " << topic << std::endl;
    }
}
